### ISSUE TITLE
Predictive performance

### ISSUE THREAD

**Issue by muhlbach** [jefferythewind/warpgbm#1]
@jefferythewind  love this and I'm all in for speed, but do you have any benchmarks comparing your implementation to XGBoost, CatBoost, etc. in terms of predictive ability? Would love to get a feeling of the improvements this library brings.

**Comment by jefferythewind** [comment]
Thanks for interest! Generally speaking GBDT is an established algorithm. LightGBM and XGBoost implement the histogram approach. I believe that CatBoost does as well but they handle integer input and categorical inputs in a different way. This package, WarpGBM follows the histogram approach. You should find performance very similar to LightGBM and XGBoost for the same parameter settings. On small datasets and shallow forests, where run times are only 1 or 2 seconds, you may not notice much of a speed difference. For millions of data points and thousands of features, you should find WarpGBM beating consistently in time. This week I will make sure to publish a more thorough benchmark. Development has been rapid. Benchmarks from last week are now outdated. And there are many more speed improvements coming down the pipeline. Finally, we will differentiate our project from the field by specializing in domain adaptation by implementing Era Splitting. Stay Tuned.

**Comment by jefferythewind** [comment]
We just updated the public README with a first based comparison of XGBoost, CatBoost, and LightGBM. It shows speed metrics and in-sample performance.

**Comment by jefferythewind** [comment]
Hi @muhlbach, the algo is the standard gradient boosting decision tree algo (GBDT). Mostly the new libraries implement what is called "the histogram method", which we implement too. What differentiates us from LightGBM and XGBoost is just that we're trying to optimize implementing the algorithm on GPUs. 

A good reference for decision trees: Classification and Regression Trees (L. Breiman) https://books.google.ca/books/about/Classification_and_Regression_Trees.html?id=JwQx-WOmSyQC&redir_esc=y

A regression tree is the base learner of the boosting algorithm. Then gradient boosting is well explained in Friedman: Greedy function approximation: A gradient boosting machine. https://www.jstor.org/stable/2699986

Then the histogram method is documented in the XGBoost paper in section 3.2: Approximate Algorithm. https://arxiv.org/pdf/1603.02754

So there is really nothing new here besides the exact code implementation, the cuda kernels. I put up a public kaggle notebook about a recent competition, just showing this single WarpGBM model can get a decent score. https://www.kaggle.com/code/jefferythewind/warpgbm-0-1-23-w-eval

So this shows we are definitely capturing the algorithm somewhat correctly. 

What is cooler is what comes next. We will implement some new ideas from the field of OOD ( out of distribution generalization ), which I've already verified work well for finding "invariant signals". Here is the paper: https://arxiv.org/abs/2309.14496

**Comment by muhlbach** [comment]
Very cool, really enjoyed reading the Era Splitting paper!
